File: ui/src/types/peptide.ts
// src/types/peptide.ts
//
// CONTRACT SOURCE: backend/schemas/api_models.py (PeptideRow)
// This file must stay in sync with the backend Pydantic model.
// Run `make contract-check` to verify alignment.
//
// SSW Semantics (LOCKED):
//   sswPrediction: 1 = positive, -1 = negative, 0 = uncertain, null = missing
// Provider Status (LOCKED):
//   providerStatus.<provider>.status: "AVAILABLE" | "UNAVAILABLE" | "PARTIAL" | "OFF"
//

// ----- Segments -----
// You previously used an object-style Segment. Keep it.
export type Segment = {
  start: number;
  end: number;
  score?: number;
};

// Also allow tuple segments (some mappers/components prefer this form).
export type SegmentTuple = [number, number];

// ----- Prediction enums -----
// SSWPrediction semantic values:
// -1 = predicted NOT to undergo structural switch (valid prediction)
//  0 = uncertain/unknown
//  1 = predicted to undergo structural switch
// null = no prediction available (provider didn't run or failed)
export type SSWPrediction = -1 | 0 | 1 | null;

// ----- Provider Status -----
// Single provider status shape
export type SingleProviderStatus = {
  enabled?: boolean;
  requested?: boolean;
  ran?: boolean;
  status: 'OFF' | 'UNAVAILABLE' | 'PARTIAL' | 'AVAILABLE';
  reason?: string | null;
  stats?: {
    requested: number;
    parsed_ok: number;
    parsed_bad: number;
  };
};

export type ProviderStatus = {
  tango?: SingleProviderStatus;
  s4pred?: SingleProviderStatus;
};

// ----- Core Peptide model used across the UI -----
// NOTE: Fields that can be `null` from the backend are typed as `| null`.
// `null` means "data is missing/unavailable" (provider didn't run or failed).
// `undefined` means "field not present in response".
// Actual zero values (0, 0.0) are valid data and should not be confused with null.
export type Peptide = {
  // identity & metadata (required)
  id: string;
  name?: string | null;
  species?: string | null;

  // UniProt metadata (optional, only present for UniProt-sourced data)
  geneName?: string | null;
  proteinFunction?: string | null;
  annotationScore?: number | null;

  // sequence (required)
  sequence: string;
  length: number | null;  // null if not computed

  // basic biophysics (can be null if not computed)
  hydrophobicity: number | null;
  muH?: number | null;  // hydrophobic moment if present
  charge: number | null;

  // Tango / SSW (Secondary Structure Switch)
  // sswPrediction can be -1/0/1 (valid predictions) or null (no prediction available)
  sswPrediction: SSWPrediction;  // from "SSW prediction"
  sswScore?: number | null;      // "SSW score" - null if TANGO unavailable
  sswDiff?: number | null;       // "SSW diff" - null if TANGO unavailable
  sswHelixPct?: number | null;   // "SSW helix percentage" - null if TANGO unavailable
  sswBetaPct?: number | null;    // "SSW beta percentage" - null if TANGO unavailable

  // FF-Helix (from auxiliary - always computed locally)
  ffHelixPercent?: number | null;                            // "FF-Helix %"
  ffHelixFragments?: Array<SegmentTuple> | Segment[];        // "FF Helix fragments"

  // FF flags (database-level binary classification from reference implementation)
  // Thresholds verified against Peleg's reference dataset (2026-02-26). Fallback: H=0.417, uH=0.388.
  ffHelixFlag?: number | null;    // 1 (candidate), -1 (not candidate), null (no data). S4PRED-based.
  ffHelixScore?: number | null;   // helix_uH + helix_score
  ffSswFlag?: number | null;      // 1 (candidate), -1 (not candidate), null (no data)
  ffSswScore?: number | null;     // Hydrophobicity + Beta_uH + Full_length_uH + SSW_prediction

  // ----- Unified secondary-structure percentages used by Results table -----
  // These are filled by the mapper from S4PRED if present, otherwise Tango fallback.
  helixPercent?: number | null;   // preferred: S4PRED helix %; fallback: "SSW helix percentage"
  betaPercent?: number | null;    // preferred: S4PRED beta %;  fallback: "SSW beta percentage"

  // --- Optional Tango per-residue curves
  tango?: {
    agg?: number[];   // Aggregation curve
    beta?: number[];  // Beta curve
    helix?: number[]; // Helix curve
    turn?: number[];  // Turn curve
  };

  // Canonical Tango summary fields (from backend, preferred over computing from extras)
  // These provide a single source of truth for UI display
  tangoHasData?: boolean;        // True if any Tango curves are available
  tangoAggMax?: number | null;   // Max value of Aggregation curve (risk indicator)
  tangoBetaMax?: number | null;  // Max value of Beta curve
  tangoHelixMax?: number | null; // Max value of Helix curve

  // ----- S4PRED secondary structure predictions -----
  // From backend/s4pred.py - matches reference 260120_Alpha_and_SSW_FF_Predictor/s4pred.py
  // Helix predictions
  s4predHelixPrediction?: number | null;              // -1 no helix, 1 helix found (alias: 'Helix prediction (S4PRED)')
  s4predHelixFragments?: Array<SegmentTuple> | null;  // segment tuples (alias: 'Helix fragments (S4PRED)')
  s4predHelixScore?: number | null;                   // avg helix score (alias: 'Helix score (S4PRED)')
  s4predHelixPercent?: number | null;                 // helix percentage 0-100 (alias: 'Helix percentage (S4PRED)')
  // SSW (Secondary Structure Switch) predictions
  s4predSswPrediction?: number | null;                // -1/1 (alias: 'SSW prediction (S4PRED)')
  s4predSswFragments?: Array<SegmentTuple> | null;    // SSW segment tuples (alias: 'SSW fragments (S4PRED)')
  s4predSswScore?: number | null;                     // SSW score (alias: 'SSW score (S4PRED)')
  s4predSswDiff?: number | null;                      // SSW diff (alias: 'SSW diff (S4PRED)')
  s4predSswHelixPercent?: number | null;              // SSW helix % (alias: 'SSW helix percentage (S4PRED)')
  s4predSswBetaPercent?: number | null;               // SSW beta % (alias: 'SSW beta percentage (S4PRED)')
  s4predSswPercent?: number | null;                   // SSW overlap % (alias: 'SSW percentage (S4PRED)')
  s4predHasData?: boolean;                            // True if S4PRED ran and produced results
  // S4PRED per-residue curves (for PeptideDetail view)
  s4pred?: {
    pH?: number[];    // per-residue P(helix)
    pE?: number[];    // per-residue P(beta)
    pC?: number[];    // per-residue P(coil)
    ssPrediction?: string[];  // per-residue SS prediction ('C', 'H', 'E')
    helixSegments?: Array<[number, number]>;  // Helix segment tuples
    betaSegments?: Array<[number, number]>;   // Beta segment tuples (from ssw fragments)
  };

  // Provider status (Principle B: mandatory provider status)
  // Row-level provider status (from row.providerStatus or row.provider_status)
  // Passed through from backend without modification
  providerStatus?: ProviderStatus;

  // passthrough for any extras
  extra?: Record<string, any>;
};

// ----- Column mapping (user remap support) -----
export type ColumnMapping = {
  entry?: string;
  accession?: string;
  sequence?: string;
  length?: string;
  hydrophobicity?: string;
  hydrophobic_moment?: string;
  charge?: string;

  // SSW (Secondary Structure Switch)
  ssw_prediction?: string;   // maps to Peptide.sswPrediction
  ssw_score?: string;
  ssw_diff?: string;
  ssw_helix_percentage?: string;
  ssw_beta_percentage?: string;

  // FF-Helix
  ff_helix_percent?: string;       // maps to Peptide.ffHelixPercent
  ff_helix_fragments?: string;

  // ----- Optional direct mapping for unified columns -----
  helix_percent?: string;          // maps to Peptide.helixPercent
  beta_percent?: string;           // maps to Peptide.betaPercent

  // metadata
  species?: string;
  notes?: string;
  name?: string;
};

// ----- Dataset-level stats for KPI cards -----
export type DatasetStats = {
  totalPeptides: number;
  sswPositivePercent: number | null; // null when TANGO unavailable
  meanHydrophobicity: number;
  meanCharge: number;
  meanMuH: number | null; // null when no μH data available
  meanFFHelixPercent: number | null; // null when no FF-Helix data available
  meanLength: number;

  meanS4predHelixPercent: number | null; // null when no S4PRED data available

  // FF candidate percentages
  ffHelixCandidatePercent?: number | null; // % of peptides with ffHelixFlag === 1
  ffSswCandidatePercent?: number | null;   // % of peptides with ffSswFlag === 1 (gated on TANGO)

  // availability counts for better UI display
  s4predAvailable?: number;
  ffHelixAvailable?: number;
  sswAvailable?: number;
  aggHotspotPercent?: number | null;
};

// ----- Raw CSV load shape -----
export type ParsedCSVData = {
  headers: string[];
  rows: Record<string, any>[];
  fileName: string;
  rowCount: number;
};

// ----- A strongly-typed “row” when exporting/saving tables -----
// Keep previous fields and add optional ones you’re now using.
export type PeptideRow = {
  Entry: string;
  Sequence: string;
  Length: number;
  Hydrophobicity: number;
  Charge: number;
  "Full length uH": number;

  // FF flags (legacy names kept)
  "FF-Helix (Jpred)": number;                 // 1 or -1
  "FF-Secondary structure switch": number;    // 1 or -1

  // Optional richer outputs (all optional to avoid breaking old sheets)
  "FF-Helix %"? : number;                     // from auxiliary calc
  "FF Helix fragments"?: Array<SegmentTuple> | any[];

  // Tango / SSW
  "SSW prediction"?: number;                  // -1/0/1
  "SSW score"?: number;
  "SSW diff"?: number;
  "SSW helix percentage"?: number;
  "SSW beta percentage"?: number;

  // JPred
  "Helix fragments (Jpred)"?: Array<SegmentTuple> | any[];
  "Helix score (Jpred)"?: number;

  // Metadata
  "Protein name"?: string;
  "Organism"?: string;
};

// ----- Threshold Configuration -----
export type ThresholdConfig = {
  mode: "default" | "recommended" | "custom";
  custom?: Record<string, any>;  // Custom threshold values when mode is "custom"
  version: string;  // Config schema version
};

// ----- Backend-provided metadata for pills at top-right -----
export type DatasetMetadata = {
  runId?: string;
  traceId?: string;
  inputsHash?: string;
  configHash?: string;
  use_tango?: boolean;
  use_s4pred?: boolean;
  ssw_rows?: number;
  valid_seq_rows?: number;
  thresholds?: {
    muHCutoff: number;
    hydroCutoff: number;
  };
  thresholdConfigRequested?: ThresholdConfig | null;
  thresholdConfigResolved?: ThresholdConfig | null;
  providerStatusSummary?: any;  // Provider status summary from backend
  provider_status?: {
    tango?: SingleProviderStatus;
    s4pred?: SingleProviderStatus;
  };
  // UniProt-specific metadata (only present for UniProt-sourced data)
  source?: "uniprot_api" | null;
  query?: string | null;
  url?: string | null;
  size_returned?: number | null;
  total_available?: number | null;
};

File: backend/schemas/peptide.py
from typing import Optional, List, Any, Dict
from pydantic import BaseModel, Field, field_validator
import pandas as pd
import math
from .provider_status import PeptideProviderStatus


def _snake_to_camel(s: str) -> str:
    parts = s.split('_')
    return parts[0] + ''.join(p.title() for p in parts[1:])


class PeptideSchema(BaseModel):
    # identity & metadata (snake_case) — aliases are the exact CSV header strings
    entry: str = Field(..., alias="Entry")
    name: Optional[str] = Field(None, alias="Protein name")
    species: Optional[str] = Field(None, alias="Organism")
    
    @classmethod
    def parse_obj(cls, obj):
        """Override parse_obj to handle NaN values in optional fields (strings, floats, ints)"""
        # Convert NaN/None to None for optional fields
        if isinstance(obj, dict):
            obj = obj.copy()
            import math
            # Handle aliases (CSV column names) and field names
            field_mappings = {
                "name": ["Protein name", "name"],
                "species": ["Organism", "species"]
            }
            for field_name, aliases in field_mappings.items():
                for key in aliases:
                    if key in obj:
                        val = obj[key]
                        # Handle pandas NaN, numpy NaN, Python None, or string "nan"
                        if val is None:
                            obj[key] = None
                        elif isinstance(val, float):
                            if pd.isna(val) or str(val).lower() == "nan" or math.isnan(val) or not math.isfinite(val):
                                obj[key] = None
                        elif isinstance(val, str) and val.lower() == "nan":
                            obj[key] = None
            
            # Sanitize SSW-related fields: convert NaN/inf to None
            ssw_fields = ["SSW prediction", "SSW score", "SSW diff", "SSW helix percentage", "SSW beta percentage"]
            for field in ssw_fields:
                if field in obj:
                    val = obj[field]
                    if val is None:
                        obj[field] = None
                    elif isinstance(val, float):
                        if pd.isna(val) or math.isnan(val) or not math.isfinite(val):
                            obj[field] = None
                    elif isinstance(val, (int, float)) and math.isnan(float(val)):
                        obj[field] = None
        return super().parse_obj(obj)

    # sequence
    sequence: str = Field(..., alias="Sequence")
    length: Optional[int] = Field(None, alias="Length")

    # basic biophysics
    hydrophobicity: Optional[float] = Field(None, alias="Hydrophobicity")
    charge: Optional[float] = Field(None, alias="Charge")
    mu_h: Optional[float] = Field(None, alias="Full length uH")

    # SSW / Chameleon (exact CSV headers)
    ssw_prediction: Optional[int] = Field(None, alias="SSW prediction")
    ssw_score: Optional[float] = Field(None, alias="SSW score")
    ssw_diff: Optional[float] = Field(None, alias="SSW diff")
    ssw_helix_percentage: Optional[float] = Field(None, alias="SSW helix percentage")
    ssw_beta_percentage: Optional[float] = Field(None, alias="SSW beta percentage")

    @field_validator('ssw_score', 'ssw_diff', 'ssw_helix_percentage', 'ssw_beta_percentage', 'ff_helix_percent', 'hydrophobicity', 'charge', 'mu_h', mode='before')
    @classmethod
    def coerce_nan_to_none(cls, v: Any) -> Optional[float]:
        """Coerce NaN/inf to None for Optional[float] fields."""
        if v is None:
            return None
        if isinstance(v, float):
            if math.isnan(v) or not math.isfinite(v):
                return None
        if isinstance(v, (int, float)):
            try:
                fv = float(v)
                if math.isnan(fv) or not math.isfinite(fv):
                    return None
                return fv
            except (ValueError, TypeError):
                return None
        return v

    @field_validator('ssw_prediction', 'length', mode='before')
    @classmethod
    def coerce_nan_to_none_int(cls, v: Any) -> Optional[int]:
        """Coerce NaN/inf to None for Optional[int] fields."""
        if v is None:
            return None
        if isinstance(v, float):
            if math.isnan(v) or not math.isfinite(v):
                return None
        if isinstance(v, (int, float)):
            try:
                iv = int(v)
                if isinstance(v, float) and (math.isnan(v) or not math.isfinite(v)):
                    return None
                return iv
            except (ValueError, TypeError):
                return None
        return v

    # FF-Helix
    ff_helix_percent: Optional[float] = Field(None, alias="FF-Helix %")
    ff_helix_fragments: Optional[List[Any]] = Field(None, alias="FF Helix fragments")
    
    # Provider status (Principle B: mandatory provider status)
    # Note: provider_status is NOT in CSV/DataFrame - it's added during normalization
    provider_status: Optional[PeptideProviderStatus] = Field(None)

    class Config:
        allow_population_by_field_name = True
        extra = "allow"

    def to_camel_dict(self) -> dict:
        """
        Convert validated model to camelCase keys used by the frontend types.
        Keeps base names identical (e.g. ff_helix_percent -> ffHelixPercent).
        
        Canonical fields:
        - sswPrediction: -1/0/1 classification (from ssw_prediction)
        - sswHelixPercentage: numeric helix percentage (from ssw_helix_percentage)
        - sswBetaPercentage: numeric beta percentage (from ssw_beta_percentage)
        """
        d = self.dict(by_alias=False, exclude_none=True)
        out = {}
        for k, v in d.items():
            # explicit exceptions that should map to different frontend key names
            if k == "entry":
                out["id"] = v
                continue
            if k == "mu_h":
                out["muH"] = v
                continue
            if k == "ssw_prediction":
                # Canonical field: sswPrediction (-1/0/1 classification)
                out["sswPrediction"] = v
                # Backward compatibility: also include chameleonPrediction (deprecated)
                out["chameleonPrediction"] = v
                continue
            
            if k == "provider_status":
                # Convert PeptideProviderStatus to dict (camelCase keys)
                if v is not None:
                    if hasattr(v, "dict"):
                        provider_dict = v.dict()
                        # Convert nested keys to camelCase if needed
                        out["providerStatus"] = provider_dict
                    else:
                        out["providerStatus"] = v
                continue

            # generic snake_case -> camelCase
            out[_snake_to_camel(k)] = v
        return out